src/lib/comms/sol-lwm2m-server.c

```diff
@@ -802,33 +812,87 @@ observer_entry_del_monitor(struct observer_entry *entry,
 }
 
 SOL_API struct sol_lwm2m_server *
-sol_lwm2m_server_new(uint16_t port)
+sol_lwm2m_server_new(uint16_t coap_port, uint16_t dtls_port,
+    enum sol_lwm2m_security_mode sec_mode, struct sol_vector *known_psks)
 {
     struct sol_lwm2m_server *server;
-    struct sol_network_link_addr servaddr = { .family = SOL_NETWORK_FAMILY_INET6,
-                                              .port = port };
+    struct sol_network_link_addr servaddr_coap = { .family = SOL_NETWORK_FAMILY_INET6,
+                                                   .port = coap_port };
+    struct sol_network_link_addr servaddr_dtls = { .family = SOL_NETWORK_FAMILY_INET6,
+                                                   .port = dtls_port };
     int r;
+    const char *sec_mode_str;
 
     SOL_LOG_INTERNAL_INIT_ONCE;
 
+    switch (sec_mode) {
+    case SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY:
+        SOL_NULL_CHECK(known_psks, NULL);
+        sec_mode_str = "Pre-Shared Key";
+        break;
+    case SOL_LWM2M_SECURITY_MODE_RAW_PUBLIC_KEY:
+        sec_mode_str = "Raw Public Key";
+        SOL_WRN("Raw Public Key security mode is not supported yet.");
+        return NULL;
+    case SOL_LWM2M_SECURITY_MODE_CERTIFICATE:
+        sec_mode_str = "Certificate";
+        SOL_WRN("Certificate security mode is not supported yet.");
+        return NULL;
+    case SOL_LWM2M_SECURITY_MODE_NO_SEC:
+        sec_mode_str = "NoSec";
+        SOL_DBG("Using NoSec Security Mode (No DTLS).");
+        break;
+    default:
+        SOL_WRN("Unknown DTLS Security Mode: %d", sec_mode);
+        return NULL;
+    }
+
     server = calloc(1, sizeof(struct sol_lwm2m_server));
     SOL_NULL_CHECK(server, NULL);
 
-    server->coap = sol_coap_server_new(&servaddr, false);
+    server->coap = sol_coap_server_new(&servaddr_coap, false);
     SOL_NULL_CHECK_GOTO(server->coap, err_coap);
 
+    if (sec_mode != SOL_LWM2M_SECURITY_MODE_NO_SEC) {
+        if (sec_mode == SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY)
+            server->known_psks = known_psks;
```

Comment:
Hmm, that makes sense.

`server->known_psks` would need to become `sol_vector` instead of `sol_vector *` then, and store elements of type `sol_lwm2m_security_psk` (so I can do
```C
uint16_t i;
struct sol_lwm2m_security_psk *psk, *given_psk;

sol_vector_init(&server->known_psks, sizeof(sol_lwm2m_security_psk));

SOL_VECTOR_FOREACH_IDX (known_psks, given_psk, i) {
    psk = sol_vector_append(server->known_psks);
    SOL_NULL_CHECK_GOTO(psk, err_psk); //TODO: add new label at the end
    psk->id = sol_blob_ref(given_psk->id);
    psk->key = sol_blob_ref(given_psk->key);
}
```
, right?

I'm not sure about adding references to the sensitive data (PSK ID and KEY) like this, though... Let's discuss this one offline as well.